### PR TITLE
new HashedSpectrum data structure for fast MSSpectrum access

### PR DESCRIPTION
**Pull request by lars20070**
Accessing a peak with a particular m/z in `MSSpectrum` can be slow. One has to iterate through the entire spectrum in order to find a specific m/z.

`HashedSpectrum` is an auxiliary data structure for the quick access of `MSSpectrum`. The spectrum is divided into a number of intervals (aka bins). Each of them contains two pointers: one to the first and one to the last peak in the interval. Given a specific m/z, one can now quickly jump into the neighbourhood i.e. bin and then search for the peak with the closest m/z.

### CODE REVIEW

File: src/openms/include/OpenMS/COMPARISON/SPECTRA/HashedSpectrum.h [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+ *
+ * @see MSSpectrum<Peak1D>
+ */
+class OPENMS_DLLAPI HashedSpectrum
+{
+  public:
+  
+    /**
+     * @brief m/z interval containing pointers to the first and last data points
+     * If there are no data points in this interval, both pointers are null.
```
Comment (timosachsenberg) [inline src/openms/include/OpenMS/COMPARISON/SPECTRA/HashedSpectrum.h]: why pointer and not iterator? why null and not .end()

File: src/openms/include/OpenMS/COMPARISON/SPECTRA/HashedSpectrum.h [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+
+    /**
+     * @brief destructor
+     */
+    ~HashedSpectrum();
+
+    /**
+     * @brief returns the m/z bin size
+     */
+    double getMzBin() const;
```
Comment (timosachsenberg) [inline src/openms/include/OpenMS/COMPARISON/SPECTRA/HashedSpectrum.h]: whats the difference between bin size and tolerance?

File: src/openms/include/OpenMS/COMPARISON/SPECTRA/HashedSpectrum.h [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+
+    /**
+     * @brief return index of the interval containing m/z
+     */
+    int getIndex_(double mz);
+
+    /**
+     * @brief return distance between two m/z in either Th or ppm (depending on mz_unit_ppm_)
+     */
+    double getDistance_(double mz1, double mz2);
```
Comment (timosachsenberg) [inline src/openms/include/OpenMS/COMPARISON/SPECTRA/HashedSpectrum.h]: const static?

File: src/openms/source/COMPARISON/SPECTRA/HashedSpectrum.cpp [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+{
+  HashedSpectrum::HashedSpectrum(MSSpectrum<Peak1D>& raw_spectrum, double mz_bin, double mz_tolerance, bool mz_unit_ppm)
+  : mz_bin_(mz_bin), mz_tolerance_(mz_tolerance), mz_unit_ppm_(mz_unit_ppm)
+  {
+    if (raw_spectrum.size() <= 2)
+    {
+      throw Exception::InvalidSize(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION, raw_spectrum.size());
+    }
+ 
+    mz_min_ = raw_spectrum.begin()->getMZ();
```
Comment (timosachsenberg) [inline src/openms/source/COMPARISON/SPECTRA/HashedSpectrum.cpp]: test if spectrum is sorted

File: src/openms/include/OpenMS/COMPARISON/SPECTRA/HashedSpectrum.h [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+     */
+    bool getMzUnitPpm() const;
+    
+    /**
+     * @brief returns pointer to the peak closest to m/z
+     * 
+     * @param mz    m/z position
+     * @return pointer to the peak closest to mz or NULL if difference between mz and closest peak m/z exceeds mz_tolerance_
+     */
+    MSSpectrum<Peak1D>::pointer getPeak(double mz);
```
Comment (timosachsenberg) [inline src/openms/include/OpenMS/COMPARISON/SPECTRA/HashedSpectrum.h]: maybe a different name? we already have getNearestPeak() why not use it as well?

File: src/openms/include/OpenMS/COMPARISON/SPECTRA/HashedSpectrum.h (hunk 1 above)
Comment (lars20070) [inline src/openms/include/OpenMS/COMPARISON/SPECTRA/HashedSpectrum.h]: I did not know what to use as `NULL` for iterators.
I will replace all pointers by iterators.

File: src/openms/source/COMPARISON/SPECTRA/HashedSpectrum.cpp [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+#include <algorithm>
+
+using namespace std;
+
+namespace OpenMS
+{
+  HashedSpectrum::HashedSpectrum(MSSpectrum<Peak1D>& raw_spectrum, double mz_bin, double mz_tolerance, bool mz_unit_ppm)
+  : mz_bin_(mz_bin), mz_tolerance_(mz_tolerance), mz_unit_ppm_(mz_unit_ppm)
+  {
+    if (raw_spectrum.size() <= 2)
```
Comment (timosachsenberg) [inline src/openms/source/COMPARISON/SPECTRA/HashedSpectrum.cpp]: why is this an invalid size? should work even with empty spectra!

File: src/openms/source/COMPARISON/SPECTRA/HashedSpectrum.cpp [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+  
+      if (index > last_index)
+      {
+        // close previous interval
+        bins_.push_back(bin);
+   
+        // add empty interval
+        bin.first = NULL;
+        bin.last = NULL;
+        for (int i=0; i < (index - last_index -1); ++i)
```
Comment (timosachsenberg) [inline src/openms/source/COMPARISON/SPECTRA/HashedSpectrum.cpp]: what does this code do?

File: src/openms/source/COMPARISON/SPECTRA/HashedSpectrum.cpp [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+  : mz_bin_(mz_bin), mz_tolerance_(mz_tolerance), mz_unit_ppm_(mz_unit_ppm)
+  {
+    if (raw_spectrum.size() <= 2)
+    {
+      throw Exception::InvalidSize(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION, raw_spectrum.size());
+    }
+ 
+    mz_min_ = raw_spectrum.begin()->getMZ();
+      
+    std::vector<double> mz;
```
Comment (timosachsenberg) [inline src/openms/source/COMPARISON/SPECTRA/HashedSpectrum.cpp]: are these vectors needed?

File: src/openms/source/COMPARISON/SPECTRA/HashedSpectrum.cpp (hunk 7 above)
Comment (timosachsenberg) [inline src/openms/source/COMPARISON/SPECTRA/HashedSpectrum.cpp]: I think I know what it is doing but could a line of comment would help

File: src/openms/include/OpenMS/COMPARISON/SPECTRA/HashedSpectrum.h (hunk 2 above)
Comment (lars20070) [inline src/openms/include/OpenMS/COMPARISON/SPECTRA/HashedSpectrum.h]: Bin size aka interval size is the m/z range associated with each each hash index. Used in the constructor.

Tolerance is the maximum difference between the mz of interest and the mz of a peak that will be returned. Typically smaller than bin size. Used in `getPeak()`.

https://github.com/OpenMS/OpenMS/pull/2377/files#diff-91a1b4a0d6dfa5b0bbbef40227ccea7aR75

File: src/openms/include/OpenMS/COMPARISON/SPECTRA/HashedSpectrum.h [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+    
+    /**
+     * @brief constructor taking an MSSpectrum
+     * 
+     * @param raw_spectrum    raw MS1 spectrum to be indexed
+     * @param mz_bin    size of the m/z bins aka intervals
+     * @param mz_tolerance    maximal difference between m/z of interest and m/z of a valid peak
+     * @param mz_unit_ppm    unit for mz_bin and mz_tolerance
+     */
+    HashedSpectrum(MSSpectrum<Peak1D>& raw_spectrum, double mz_bin, double mz_tolerance, bool mz_unit_ppm);
```
Comment (timosachsenberg) [inline src/openms/include/OpenMS/COMPARISON/SPECTRA/HashedSpectrum.h]: why mz_bin and mz_tolerance parameter? wouldn't it make sense to choose mz_bin depending on mz_tolerance? Or would this result in too many bins?

Comment (timosachsenberg) [inline src/openms/include/OpenMS/COMPARISON/SPECTRA/HashedSpectrum.h]: Looking at the code below it seems mz_tolerance does not influence the number of bins so I guess it doesn't need to be set here?

File: src/openms/source/COMPARISON/SPECTRA/HashedSpectrum.cpp [hunk 10]
```diff
@@ -0,0 +1,236 @@
...
+     return (int) floor((mz - mz_min_)/mz_bin_);
+    }
+  }
+  
+  double HashedSpectrum::getDistance_(double mz1, double mz2)
+  {
+    if (mz_unit_ppm_)
+    {
+      // m/z tolerance in ppm
+     return std::abs(mz1 - mz2)/mz1*1.0e+6;
```
Comment (timosachsenberg) [inline src/openms/source/COMPARISON/SPECTRA/HashedSpectrum.cpp]: please check throughout the code if you can replace your calculations with functions from https://github.com/OpenMS/OpenMS/blob/develop/src/openms/include/OpenMS/MATH/MISC/MathFunctions.h

File: src/openms/source/COMPARISON/SPECTRA/HashedSpectrum.cpp (hunk 7 above)
Comment (lars20070) [inline src/openms/source/COMPARISON/SPECTRA/HashedSpectrum.cpp]: Adds bins that do not contain any data points. In this case both pointer (soon iterators) are `NULL`.

File: src/tests/class_tests/openms/source/HashedSpectrum_test.cpp [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+#include <OpenMS/KERNEL/MSExperiment.h>
+#include <OpenMS/FORMAT/MzMLFile.h>
+
+using namespace OpenMS;
+
+START_TEST(MultiplexFiltering, "$Id$")
+
+// read data
+MSExperiment<Peak1D> exp;
+MzMLFile().load(OPENMS_GET_TEST_DATA_PATH("MultiplexFiltering.mzML"), exp);
```
Comment (timosachsenberg) [inline src/tests/class_tests/openms/source/HashedSpectrum_test.cpp]: could you please add a test that creates the values in the code. This might be data structure of general use and more extensive testing will probably be required (especially of corner cases)

File: src/openms/include/OpenMS/COMPARISON/SPECTRA/HashedSpectrum.h (hunk 1 above)
Comment (timosachsenberg) [inline src/openms/include/OpenMS/COMPARISON/SPECTRA/HashedSpectrum.h]: maybe I missed something but either you have several points in the bin: two iterators pointing to first and last.
one point in the bin; two iterators - first and last both pointing to the single point
no point: two iterators - both pointing to spectrum.end()

File: src/openms/include/OpenMS/COMPARISON/SPECTRA/HashedSpectrum.h (hunk 2 above)
Comment (timosachsenberg) [inline src/openms/include/OpenMS/COMPARISON/SPECTRA/HashedSpectrum.h]: so why set it in the constructor?

File: src/openms/include/OpenMS/COMPARISON/SPECTRA/HashedSpectrum.h (hunk 1 above)
Comment (lars20070) [inline src/openms/include/OpenMS/COMPARISON/SPECTRA/HashedSpectrum.h]: Correct. That are the three cases.

I will use `spectrum.end()` instead of `NULL`. Good point.

**Comment by timosachsenberg** [comment]
@lars20070 I think this is a very valuable addition. As this class is of general interest other OpenMS developers might also post their comments and suggestions.

File: src/openms/include/OpenMS/COMPARISON/SPECTRA/HashedSpectrum.h (hunk 2 above)
Comment (lars20070) [inline src/openms/include/OpenMS/COMPARISON/SPECTRA/HashedSpectrum.h]: I can pass the tolerance in `getPeak()`. It is the more logical place and more flexible.

But in my use case the tolerance will always be the same. So I pass the tolerance just once in the constructor.

Comment (timosachsenberg) [inline src/openms/include/OpenMS/COMPARISON/SPECTRA/HashedSpectrum.h]: please add it to the function that uses it

**Comment by hroest** [comment]
We had the same problem in OpenSWATH, however we solved it in a different way. Basically, we sorted the spectra and then used binary search to find peaks in `log(n)` time. Did you try that? Depending on your bin size this may even be faster (e.g. you will have `O(1) + O(k)` where `k` is the average peaks per bin whereas binary search is O(log(n))` so depending on k the new method may be faster or slower).

Just a comment on what we did and it does not use extra memory or extra computation except an initial sorting by m/z (which most spectra already are). Basically, we just used `std::lower_bound` on our spectra, see https://github.com/OpenMS/OpenMS/blob/develop/src/openswathalgo/source/ANALYSIS/OPENSWATH/OPENSWATHALGO/DATAACCESS/SpectrumHelpers.cpp#L95

**Comment by cbielow** [comment]
Along the lines of @hroest:

@lars20070 have you tried
[Size 	findNearest (CoordinateType mz) const](http://ftp.mi.fu-berlin.de/pub/OpenMS/release-documentation/html/classOpenMS_1_1MSSpectrum.html#a830e58a22131827d85007a70e3069a5f)
There are three overloads available. It might even be faster and there is no need for an additional datastructure.

**Comment by timosachsenberg** [comment]
I could imagine that this is faster for small bin sizes where most have 0 or 1 peak in each bin. @lars20070 I guess you did some benchmarks?

**Comment by lars20070** [comment]
@hroest Combining both approaches, i.e. doing a binary search in my m/z bins should be the fastest.

You are right, the bin size is crucial. I plan to pass this through as advanced parameter and optimize the run time for each type of data. Of course, a dynamic bin size estimation would be better.

**Comment by lars20070** [comment]
@cbielow That's a quick way of bench marking against binary search. Will give this a go. Thank you.

**Comment by lars20070** [comment]
@timosachsenberg The current implementation (peak registry) is basically a one/none index. This is quite fast, but not general enough (as discussed on Gitter).
https://github.com/OpenMS/OpenMS/blob/develop/src/openms/include/OpenMS/TRANSFORMATIONS/FEATUREFINDER/MultiplexFiltering.h#L256
Haven't done any benchmarking yet. But a conclusive benchmarking would help to avoid duplicate implementations for essentially the same problem.

File: src/openms/include/OpenMS/COMPARISON/SPECTRA/HashedSpectrum.h (hunk 1 above)
Comment (lars20070) [inline src/openms/include/OpenMS/COMPARISON/SPECTRA/HashedSpectrum.h]: @timosachsenberg  In order to replace `NULL` with `spectrum.end()` I will need to hold a reference to `spectrum`.
https://github.com/lars20070/OpenMS/blob/feature/HashedSpectrum/src/openms/source/COMPARISON/SPECTRA/HashedSpectrum.cpp#L154

Not sure if that simplifies the API. Can we chat/phone?

Comment (timosachsenberg) [inline src/openms/include/OpenMS/COMPARISON/SPECTRA/HashedSpectrum.h]: or more STL compliant:
no point: both iterators pointing to spectrum.end()
one point in the bin; two iterators - first pointing to the single point, last to next point (or spectrum.end() if there is none)
several points in the bin: first points to first point, last to one after last in interval  (or spectrum.end() if there is none)

File: src/openms/include/OpenMS/COMPARISON/SPECTRA/HashedSpectrum.h [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+	};
+    
+    /**
+     * @brief constructor taking an MSSpectrum
+     * 
+     * @param spectrum    MS spectrum to be indexed
+     * @param mz_bin    size of the m/z bins aka intervals
+     * @param mz_unit_ppm    unit for mz_bin 
+     */
+    HashedSpectrum(MSSpectrum<Peak1D>& spectrum, double mz_bin, bool mz_unit_ppm);
```
Comment (timosachsenberg) [inline src/openms/include/OpenMS/COMPARISON/SPECTRA/HashedSpectrum.h]: const MSSpectrum? and ConstIterator at most places?

File: src/openms/source/COMPARISON/SPECTRA/HashedSpectrum.cpp [hunk 13]
```diff
@@ -0,0 +1,234 @@
...
+#include <OpenMS/COMPARISON/SPECTRA/HashedSpectrum.h>
+
+#include <vector>
+#include <algorithm>
+
+using namespace std;
+
+namespace OpenMS
+{
+  HashedSpectrum::HashedSpectrum(const MSSpectrum<Peak1D>& spectrum, const double mz_bin, const bool mz_unit_ppm)
```
Comment (timosachsenberg) [inline src/openms/source/COMPARISON/SPECTRA/HashedSpectrum.cpp]: you forgot to also remove mz_unit_ppm from the constructor

File: src/openms/include/OpenMS/COMPARISON/SPECTRA/HashedSpectrum.h (hunk 1 above)
Comment (lars20070) [inline src/openms/include/OpenMS/COMPARISON/SPECTRA/HashedSpectrum.h]: @timosachsenberg Say we have a sequence of three bins `... A B C ...` where `B` is empty, and `A` and `C` are not. Does `end` in `A` point to `spectrum_.end()` or the first data point in `C`?

I find the existing convention easier to work with.
https://github.com/OpenMS/OpenMS/pull/2377/files#diff-91a1b4a0d6dfa5b0bbbef40227ccea7aR64

File: src/openms/source/COMPARISON/SPECTRA/HashedSpectrum.cpp (hunk 13 above)
Comment (lars20070) [inline src/openms/source/COMPARISON/SPECTRA/HashedSpectrum.cpp]: No. The bins can be in Th or ppm.

File: src/openms/source/COMPARISON/SPECTRA/HashedSpectrum.cpp [hunk 14]
```diff
@@ -0,0 +1,235 @@
...
+  {
+    return mz_unit_ppm_;
+  }
+  
+  int HashedSpectrum::getIndex_(const double mz) const
+  {
+    if (mz_unit_ppm_)
+    {
+      // m/z intervals in ppm
+      return (int) floor(log(mz/mz_min_)/log(1.0 + mz_bin_*1.0e-6));
```
Comment (timosachsenberg) [inline src/openms/source/COMPARISON/SPECTRA/HashedSpectrum.cpp]: can you document how this is calculated?

File: src/openms/include/OpenMS/COMPARISON/SPECTRA/HashedSpectrum.h [hunk 15]
```diff
@@ -0,0 +1,152 @@
...
+{
+  public:
+  
+    /**
+     * @brief m/z interval (aka bin) containing iterators pointing to the first and last data points
+     * Boundaries of the interval are implicitly given by mz_min, mz_bin, mz_unit_ppm_
+     * and the hash table index.
+     * 
+     * empty m/z interval:                         both begin and end point to spectrum_.end()
+     * one data point in m/z interval:             both begin and end point to this data point
```
Comment (timosachsenberg) [inline src/openms/include/OpenMS/COMPARISON/SPECTRA/HashedSpectrum.h]: is this comment still correct?

File: src/openms/include/OpenMS/COMPARISON/SPECTRA/HashedSpectrum.h [hunk 16]
```diff
@@ -0,0 +1,152 @@
...
+     */
+    bool getMzUnitPpm() const;
+    
+    /**
+     * @brief find the peak closest to m/z
+     * 
+     * @param mz    m/z position
+     * @param mz_tolerance    m/z tolerance between mz and peak
+     * @param mz_unit_ppm    unit for mz_tolerance (ppm or Th)
+     * @return pointer to the peak closest to mz or NULL if difference between mz and closest peak exceeds the tolerance
```
Comment (timosachsenberg) [inline src/openms/include/OpenMS/COMPARISON/SPECTRA/HashedSpectrum.h]: update comment

File: src/openms/include/OpenMS/COMPARISON/SPECTRA/HashedSpectrum.h (hunk 15 above)
Comment (lars20070) [inline src/openms/include/OpenMS/COMPARISON/SPECTRA/HashedSpectrum.h]: Please see comment here.
https://github.com/OpenMS/OpenMS/pull/2377#discussion_r99130875

File: src/openms/include/OpenMS/COMPARISON/SPECTRA/HashedSpectrum.h (hunk 1 above)
Comment (timosachsenberg) [inline src/openms/include/OpenMS/COMPARISON/SPECTRA/HashedSpectrum.h]: first point in C

Comment (timosachsenberg) [inline src/openms/include/OpenMS/COMPARISON/SPECTRA/HashedSpectrum.h]: The existing convention is the STL convention ;)

**Comment by lars20070** [comment]
@timosachsenberg Probably even faster with the STL convention.

File: src/openms/source/COMPARISON/SPECTRA/HashedSpectrum.cpp [hunk 17]
```diff
@@ -216,7 +212,7 @@ namespace OpenMS
     // loop over relevant peaks
     MSSpectrum<Peak1D>::ConstIterator it_mz_closest = first_peak;
     double distance_closest = getDistance_(mz, first_peak->getMZ());
-    for (MSSpectrum<Peak1D>::ConstIterator it_mz = first_peak; it_mz <= last_peak; ++it_mz)
+    for (MSSpectrum<Peak1D>::ConstIterator it_mz = first_peak; it_mz < last_peak; ++it_mz)
```
Comment (timosachsenberg) [inline src/openms/source/COMPARISON/SPECTRA/HashedSpectrum.cpp]: I think it is valid to compare with < but comparison with != would be correct

**Comment by lars20070** [comment]
@timosachsenberg `HashedSpectrum` is twice as fast as the old `findNearest()`.
https://github.com/OpenMS/OpenMS/pull/2377/commits/fb6fd8de8d84605fedf0bb15818f8c6490cc2ed4#diff-bc04feff7c609617fcf46d3568178fa6R1040

Binary search in the bins makes makes no difference, if the bin size is optimal. But by using binary search we ensure good performance even if the bin size is sub-optimal.

`ppm` bins (with rougly the same number of points per bin) do not perform better than `Th` bins. Probably because of the cost of the `log()`s.

**Comment by timosachsenberg** [comment]
hmm ok a factor of two is not really breathtaking as it also consumes more memory. In this case, it might be a better idea to first implement your ideas without the HashedSpectrum. Later you can still decide if it makes sense to switch to it as the overall gain is rather small.

**Comment by lars20070** [comment]
You are right, @timosachsenberg. One saves only a couple of jumps in the binary search with the new data structure. Not worth the code and memory.
